File: fib.php
#!/usr/bin/env php
<?php
//--------------------------------------------------------------------
// BigNum: extension=gmp
// $ php.exe fib.php [NUM]
//--------------------------------------------------------------------
function subFib(
	$num
)
{
	if($num < 0)
	{
		return;
	}

	echo "0\t0", PHP_EOL;
	if($num < 1)
	{
		return;
	}

	echo "1\t1", PHP_EOL;
	if($num < 2)
	{
		return;
	}

	$fib0 = 0;
	$fib1 = 1;
	$fib2 = 0;
	$cnt = 1;

	$BufMax = 4 * 1000;
	$Buf = [$BufMax];
	$BufIndex = 0;

	while($num > $cnt)
	{
		$fib2 = gmp_add($fib1, $fib0);
		$fib0 = $fib1;
		$fib1 = $fib2;
		++$cnt;

		$Buf[$BufIndex + 0] = $cnt;
		$Buf[$BufIndex + 1] = "\t";
		$Buf[$BufIndex + 2] = gmp_strval($fib1);
		$Buf[$BufIndex + 3] = PHP_EOL;
		$BufIndex += 4;

		if($BufIndex >= $BufMax)
		{
			echo implode("", $Buf);
			$BufIndex = 0;
		}
	}

	echo implode("", array_slice($Buf, 0, $BufIndex));
}

function main()
{
	$ARGC = $_SERVER['argc'];
	$ARGV = $_SERVER['argv'];
	$num = 0;
	if($ARGC > 1 && is_numeric($ARGV[1]))
	{
		$num = (int)$ARGV[1];
	}
	subFib($num);
}

main();
?>

File: fib.c
//------------------------------------------------------------------------------
// BigNum
// <Windows>
//   $ gcc -O3 -Wall fib.c -static -lgmp -o fib.c.exe & strip fib.c.exe
//   $ fib.c.exe [NUM]
// <Linux>
//   $ gcc -O3 -Wall ./fib.c -static -lgmp -o ./fib.c.bin ; strip ./fib.c.bin
//   $ ./fib.c.bin [NUM]
//------------------------------------------------------------------------------
#include <gmp.h>
#include <stdio.h>
#include <stdlib.h>

void
subFib(
	int num
)
{
	if(num < 0)
	{
		return;
	}

	puts("0\t0");
	if(num < 1)
	{
		return;
	}

	puts("1\t1");
	if(num < 2)
	{
		return;
	}

	mpz_t fib0, fib1, fib2;
		mpz_init_set_si(fib0, 0);
		mpz_init_set_si(fib1, 1);
		mpz_init_set_si(fib2, 0);
	size_t iCnt = 1;

	// Buf Default
	const size_t BufBase = 5000;
	// strlen(fib1) ≒ iCnt * 0.21
	const double BufDmzRate = 0.21;
	//           Average   ProcCount
	// (BufDmzRate * 0.5 * 100) => (BufDmzRate * 50)
	const double BufAddAve = BufDmzRate * 50;
	size_t BufDmz = BufBase * BufDmzRate;
	size_t BufMax = BufBase * BufAddAve;
	char *Buf = calloc((BufDmz + BufMax), sizeof(char));
	size_t BufEnd = 0;

	while(num > iCnt)
	{
		mpz_add(fib2, fib1, fib0);
		mpz_set(fib0, fib1);
		mpz_set(fib1, fib2);
		++iCnt;

		BufEnd += gmp_sprintf((Buf + BufEnd), "%zu\t%Zd\n", iCnt, fib1);

		if(BufEnd >= BufMax)
		{
			fputs(Buf, stdout);

			size_t _i1 = iCnt + BufBase;
			BufDmz = _i1 * BufDmzRate;
			BufMax = _i1 * BufAddAve;
			Buf = realloc(Buf, ((BufDmz + BufMax) * sizeof(char)));
			*Buf = 0;
			BufEnd = 0;
		}
	}

	fputs(Buf, stdout);

	free(Buf);
}

int
main(
	int ARGC,
	char* ARGV[]
)
{
	int num = 0;
	if(ARGC > 1)
	{
		num = atoi(ARGV[1]);
	}
	subFib(num);
	return 0;
}
